docs/examples/nmt/transformer.md

```diff
@@ -0,0 +1,603 @@
+# Transformer based Machine Translation Using GluonNLP
+
+In this notebook, we will show how to train Transformer and evaluate the pretrained model on WMT 2014 English-German dataset using GluonNLP. We will together go through: 1) load and process dataset, 2) define the Transformer model, 3) train and evaluate the model, and 4) use the state-of-the-art pretrained Transformer model.
+
+## Preparation 
+
+### Load MXNet and Gluon
+
+```{.python .input  n=1}
+import warnings
+warnings.filterwarnings('ignore')
+
+import time
+import random
+import os
+import io
+import logging
+import math
+import numpy as np
+import mxnet as mx
+from mxnet import gluon
+from mxnet.gluon.data import ArrayDataset, SimpleDataset
+from mxnet.gluon.data import DataLoader
+import gluonnlp.data.batchify as btf
+from gluonnlp.data import SacreMosesDetokenizer
+from gluonnlp.data import ShardedDataLoader
+from gluonnlp.data import ExpWidthBucket, FixedBucketSampler, WMT2014BPE
+from gluonnlp.model import BeamSearchScorer
+from scripts.nmt.translation import NMTModel, BeamSearchTranslator
+from scripts.nmt.transformer import get_transformer_encoder_decoder
+from scripts.nmt.loss import SoftmaxCEMaskedLoss, LabelSmoothing
+from scripts.nmt.utils import logging_config
+from scripts.nmt.bleu import _bpe_to_words, compute_bleu
+import scripts.nmt._constants as _C
+from scripts.nmt.dataset import TOY
+```
+
+### Set Environment
+
+```{.python .input  n=2}
+np.random.seed(100)
+random.seed(100)
+mx.random.seed(10000)
+ctx = mx.gpu(0)
+```
+
+### Set Hyperparameters
+
+```{.python .input}
+# parameters for dataset
+demo = True
+if not demo:
+    dataset = 'WMT2014BPE'
+else:
+    dataset = 'TOY'
+src_lang = 'en'
+tgt_lang = 'de'
+src_max_len = -1
+tgt_max_len = -1
+
+# parameters for model
+num_units = 512
+hidden_size = 2048
+dropout = 0.1
+epsilon = 0.1
+num_layers = 6
+num_heads = 8
+scaled = True
+
+# parameters for training
+optimizer = 'adam'
+epochs = 3
+batch_size = 2700
+test_batch_size = 256
+num_accumulated = 1
+lr = 2
+warmup_steps = 1
```

Comment:
This can makes stepsize too large and possibly cause NAN.

Reply:
It works fine for now, `step_size = float(loss_denom) / batch_size / 100.0`, since the `batch_size` is large. So you suggest use smaller `lr`?